README
iWar [Intelligent Wardialer] Version 0.071
By Da Beave [Jan. 2006]
beave@softwink.com

ftp://ftp.vistech.net/pub/iwar
http://www.softwink.com/iwar

** DISCLAIMER:  iWar is a phone network security auditing tool and should only
                be used for _legal_ purposes.   If you do not have permission
		to scan a block of numbers,  then don't use this tool!


Introduction:
-------------

	It'd be my guess that shortly after the first auto dialing "MODEM"
(MODulation/DEModulation) was invented,   "demon dialing" came about. 
This would have happened in the pre-"Wargames,  the movie" days.   Yes, 
"demon dialing",  more modernly known as "wardialing",  has been around 
longer than "David Lightman's" days of "wardialing" with a acoustic modem
(I never quite understood that).   Of course,  the term "wardialing" took
off after the movie "wargames",  which is okay by me,  as "wardialing"
just sounds so much cooler than "demon dialing". 

	Lately,  the traditional press would have you believe that 
"wardialing" only took place in the 80's and 90's,  and has just recently 
made a "come back".   It's been a fairly well know fact in the underground 
"hacking" community that while organizations will throw money at firewalls, 
IDS/IPS (Intrusion Detections Systems/Intrusion Prevention Systems), 
and other Internet based safe guards,   dialup modems largely go unnoticed 
and unchecked.

	Dialup modems are usually the path to least resistance in regards
to security.

	So what is "wardialing"?  The concept hasn't changed much since
and before the movie "Wargames" (1983).  A computer, or set of computers,  
dialing numbers randomly or sequentially and recording things that 
might be interesting.     Interesting things range from traditional 
carrier detection to telco related equipment.

Simple Wardialing:
------------------

	By "simple",  I'm referring to a computer,  a modem and a good ole
POTS (Plain Old Telephone Service) line.  This is sort of the "old school"
way of doing things.    The "old school" method isn't that much different
than the "new school" method,  they just use different means to get the
call out on the PSTN (Public Switched Telephone Network). 

	If this is the manner you're going to "wardialing" a block of numbers, 
it would be wise to contact your local telephone company and let them
know that you are auditing  _your_ lines (You are using iWar in a 
_LEGAL_ manner, correct?).  

	Some Telco's don't care if you're legally auditing or not,  they'll
"flag" your line either way if they notice strange activity.    Once
"flagged", they typically shut off your POTS line.

	To avoid complications from your local phone company,  there's
several things you can do. 

	First and foremost,  if the block you're scanning consists over
100 numbers,   do _not_ attempt to do it all at once.    Take your time,
and use the random dialing function (the default).   I'd advise doing
20-50 numbers at a time,  and waiting a few hours.   

	As a pre-string (-e option),  you can also add "*67" to block 
your Caller ID.  This shouldn't every be confused with ANI.   While the
remote caller will see "Caller ID Blocked",  the ANI information is still
out there.  This just prevents the person you called from hitting 
*69 and yelling at you.  *67 is a free service provided by your
telephone company.
	
	The moral?   Keep it simple.  Keep it "slow and low" (under the
radar).   This would be the same method a attacker would use on your
block,  so why not mimic them?  If possible,  don't even use the "simple" 
or "old school" method.   

Wardialing with VoIP hardware (Cisco ATA/Linksys PAP2):
------------------------------------------------------

	Wardialing over VoIP (Voice over Internet Protocol) is a better
way of doing things.    The equipment needed is a modem,  computer,  a
decent Internet connection and some sort of VoIP hardware (Cisco ATA, 
Linksys PAP2, etc).   The idea is basically the same,  but rather than
your phone call taking place over the PSTN,  you're using the a outbound
VoIP provider.  Unlike the "old school" method,  this adds a extra
layer between you and the target number being called.   For example, 
the ANI passed is that of your VoIP provider,   and not directly 
tied to you.   Another advantage is that many VoIP provider let you
"spoof" caller ID.   So,  not only will the target not be able to call
you back,  but you can supply false information.   What I typically
do is put a known fax number as my Caller ID.  This way,  if the target
decides to call me back and yell,  he'll get a fax machine.   The logic
is that the caller will believe it was some sort of fax screwup/miss dial.

	To set your hardware like a ATA to spoof your caller ID,  check
the manual/configuration of your device.

	A wonderful step is use Asterisk (http://www.asterisk.org) to
place the call.   The equipment involved then become a modem,  computer, 
a VoIP device (ATA, PAP2, etc) and a server running Asterisk.   The idea
is to have your VoIP device place the call through your Asterisk server.   
Caller ID spoofing can then be setup there.   As well,  you can record the 
audio for your scan for later review!  Here's a example of my outbound
dialing context:

exten => _8.,1,Set(CALLERID(name)=FAX)
exten => _8.,2,SetCIDNum(8505551212|a)
exten => _8.,3,monitor,wav|${EXTEN:1}
exten => _8.,4,Dial(IAX2/myusername@MyVOIPprovider/${EXTEN:1})
exten => _8.,5,Congestion

	Breaking this down,  we set the caller ID name to "FAX".   We
then set "our phone number" to "850551212".   The "monitor" statement
tells the Asterisk server to "record" the call.   To keep things simple, 
we record the call as the phone number dialed (for example,  8505551234.wav).
Once your scan is complete,  you can go back and review the .wav files
to see if you missed anything interesting.

	It should be noted that since you're using VoIP to handle your
scanning,   only use the ulaw/alaw (G.711) codecs!  Codecs like GSM and 
G729 can cause havoc on your scans!  Even more so when you scanning simply 
for carriers.   You might have noticed that iWar defaults to a very 
low speed (1200 bps).   This seems to be a reasonable speed for iWar
to work and identify carriers.   The down fall is that a few modems
won't even bother trying to connect at low speeds.  I've found this
to be fairly rare,  but it does happen.  The low rate also lets
iWar attempt to identify the remote system.   Higher speeds over
VoIP tend to not work very well,   and produce a lot of garbage over 
the connection.   If you're using VoIP,  use lower speeds when scanning.
If you're using the "old school" method,  feel free to bump up the 
rate.

	Just because you're using a VoIP provider to scan doesn't make 
it impossible to track you back.  Someone with enough time and energy
can have your VoIP provider track you back.   This shouldn't be to much
of a problem as long as your use your brain a little bit.


Wardialing _without_ VoIP Hardware:
-----------------------------------

	Imagine you're sitting in a place with a wifi hot spot,  and
you wish to do some scanning.  Or perhaps you don't own any 
VoIP hardware (Cisco ATA/Linksys, PAP2, Modem, etc.) to do scanning with.
Using iWar with IAX2 support,  you can scan with nothing more
than a Internet connection, Intra-Asterisk eXchange (IAX2) VoIP 
provider (ie - nufone, iaxcc, local Asterisk server) and a set headphone 
w/ microphone (optional)!

	iWar works the same way as normal war dialing,  but rather than
using analog modem hardware,  it used the IAX2 VoIP protocol for dialing.  
iWar will act as a _full blown_ IAX2 VoIP client.   That is,  not only 
will it dial target numbers and let you listen to the results,  but using 
your microphone you can "talk" to the remote target (if so desired)!  
In IAX2 mode,  iWar keys 1-9, * and # send there DTMF equivalents.   
With the IAX2 add functionality, I've add a additional "Pause" function.   
If you need to simply get up (ie - to use the bathroom) and want to 
pause scanning,  simply hit 'p'.  In the event you find something/someone 
interesting and wish to pause and "MARK" the number,  hit '['.   
This will not only pause the software, but will log it,  and allow you 
to talk/play with the called number!

	At this time,  I've only added IAX2 client functionality.   In
IAX2 dial mode,  it will _not_ detect remote event (for example,  "RINGING", 
carrier,  fax,  voice).   I'm currently working on the "signal processing"
code to detect such events. 

	For information about how to build iWar with IAX2,  see the 
README.IAX2 file. 


Tone Location With Modem Hardware (non-IAX2 calls):
---------------------------------------------------

	iWar uses two methods for "tone detection".  Actually,  one 
method is true "tone detection",  the other is "silence detection". 
Unfortunately,  wardialing is tied directly to the hardware (modem)
you are using.   Considering this is all serial communication,  there
is no way for your computer to "listen" to the tone and determine if
its something interesting or not.  That's left up to the modem.   So, 
if you're using a crapy modem for other than carrier detection,  then
expect crapy result (See the modem hardware section). 

	iWar uses the traditional "ToneLoc" tone detection.   Again,  
this only works for a small set of modems.  The idea is rather than 
send the tradition dial string (ATDT5551212), you send ATDT5551212w;.  
This tells the modem to dial the number and "wait" for an addition 
tone.  If a "tone" is "heard" by the modem,  it'll return back to the 
command prompt (with a "OK").  If the modem receives a "BUSY" or "VOICE",  
it is ignored.   If it received a "OK",  we know that we got some 
sort of "tone" on the called number and it is logged.  

	The downfall of this,  depending on your hardware,  is that 
it requires you to scan a block twice.  One time for carriers,  the 
second time for tones.    

	It's also not terribly accurate (See the modem hardware section!).
Some modems will record a voice or ring as a possible tone.  Some modems
don't even bother really checking and return a "OK" no matter what. 

	The second method is a "silence" search.   The idea is pretty simple.
If your modem supports remote call progression indication (for example, 
"RINGING"),   then we watch for remote rings.   In the iwar configuration
file,  there is a "remote_ring" option.   If we call a number and the
number doesn't answer after X number of rings,  we hang up (speeds up
scanning).   If we call a number,   and it rings under the amount
of remote_rings ,  without receiving a VOICE or BUSY,  we can assume
something picked up.    If it didn't make a sound or the modem didn't
detect it,  then it might be something interesting and iWar records it. 

	Normal "tone detection" is disable by default,  as many modems
don't support it.    Also,   some modems will detect tones,  but won't
detect carriers.   Other modems can do both.   To enabled tone location
(ATDT5551212w;) you can pass iWar the -a flag.  

	By default,  "silence detection" is disabled.  If your modem
supports remote call progression,  set the "remote_ring" to the amount
of rings you want to wait on.   The good thing about silence detection
is that it'll allow you to scan for silence and carriers.   You should
carefully test your modems capabilities.  For example,  the USR Couriers
will sometimes detect a remote tone as "VOICE" (ATX6 - enable all
result codes).   If this is the case,  you'll miss a lot of goodies.  
You might want to add ATX7 (enabled everything except VOICE detection).

	ATX6 with the USR Courier is great for simply scanning for carriers,
if you're not interested in silence detection.   It'll detect and move 
on to the next number a lot faster,   but you'll miss "silent answers". 

	The default iWar setting assume you're using a standard old modem.
That is,  no remote call progression and "ATDT5551212w;" will not work. 
You'll need to tweak you're configuration to work best for you (via the
modem init string).

        iWar can also scan for Fax machines as well.   That might not
sound terrible interesting,   but can lead to some useful information.
If you have a "fax modem",  simply add the "+FCLASS=1" to your init string.   
Unfortunately, if you're scanning for fax machines,   you will miss 
normal carriers.  From my understanding,  some ZyXEL modems will support 
both fax and carrier detection.   That would be rather nifty,  but 
considering I've been unsuccessful on getting my hands on one of these
modems,  I haven't been able to test this.

        If you're scan finds fax machines,  you can in turn use Minicom
to connect to them (remember,   add "+FCLASS=1" to your terminals init
string).   What you'll find one connection is that many faxes return the 
company name that owns them.  This information can be useful.


So, I've found stuff.  Now what?
--------------------------------
	
	This is pretty much up to you,  and there's so much crazy stuff
out there,  it'd be way outside the scope of this README to cover every
possible scenario.   If the number(s) found seem to have fax machines
within close range,  that might give you a idea of who owns the device.
You can also use the "backspoofing" method to determine who own the 
phone numbers.  The basic idea is to setup your Caller ID to the number
your interested in and call yourself.   The phone company's database 
should display who own the number (for more information,  Google
"backspoof").  There are also many resources on the Internet that will
assist you in identifying remote numbers you've found. 

Modem Hardware
--------------

	As stated before,   scanning always depends on the quality of
the modem.  Cheaper modems support the ATX0-4.   This means they 
can "blind dial" (don't look for dial tone, just start dialing) to 
detecting BUSY,  VOICE,  NO CARRIER,  etc.  I've found that some 
cheaper modems will detect voice as busy.  Many do not support 
ATDT5551212w; type tone detection.   They simply always give a "OK"
(return to command prompt) on tone detection regardless how the remote
end answers.   If you're looking to quality scanning,   investing in
a decent modem is a good step.   For example,   USR Couriers,  high
end ZyXEL and Multitech modems work well.  These support extended 
result codes which you can use to your advantage.   You can usually
find these modems on Ebay for a pretty good price.

	iWar does support multiple modems.   When using iWar with more
than one modem,  use the -t flag to specify the modem you want to use.
iWar will work with as many modems your system can support!

	Before e-mailing that something doesn't work,  please check 
the capabilities of your modem first.  This can be done by simply
firing up your favorite terminal software (minicom,  seyon, etc) 
and doing some test dials.   


Features:                                                                       
---------

-       Full and Normal logging:   Full logging records all possible
        events during dialing (busy signals,  no answers,  carriers,
        etc).  Normal it only records things that we might find interesting
        (carriers,  possible telco equipment).                          

-       ASCII flat file and MySQL logging:  You can log to a traditional
        ASCII flat file,   and record information into a MySQL database.

-       Dials randomly or sequentially.                                

-       Remote system identifications:   When finding a remote modem and
        connecting, iWar will remain connected and attempt to identify  
        the remote system type.   This is based off "banner detection",
        and is not 100% fool proof!                                    

-       Key stroke marking:   When actively "listening" to iWar working,
        if you hear something interesting,  you can manually "mark" it  
        by hitting a key.   You can also enter a "note" about something
        you find interesting.                                          

-       Multiple modem support,  because...  well,  hey - this is Linux.
        iWar will support as many modems you can hook up                

-       Nice "curses" based display.  This means that if you're using
        iWar from a Linux console or a VT100 based terminal,  it should
        work fine.   It's not a escape sequence kludge,  but true "curses".

-       Full control over the modem:   Unlike other 'kludges',  iWar      
        doesn't just open the modem as a typical "file".   It controls
        the baud rate,  parity,  and CTS/RTS (Hardware flow control)  
        DTR (Data terminal read).   This is important for controlling
        the modem and making it preform the way you want it to during
        scanning.  For example,  DTR hang ups                        

-       Blacklisted phone number support:  For numbers the system
        should _never_ dial.                                     


ChangeLog
[01/16/2006] - 0.071 Release, "Now with 101% more VoIP" - Beave
 
               Fixed problem with MySQL logging.   MySQL logging was broken
	       with the 0.07 release.

	       Fixed problem with "custom" keystroke marking.  It wasn't 
	       saving the data.   It also wasn't inserting the data into
	       the database correctly.

	       Fixed a "hang" problem during IAX2 mode when a MySQL error
	       happened. 

	       Changed callerid handling a bit.   Do _not_ run if the 
	       user sets the caller ID to "911".  Also,  if 
	       iax2_callerid_number is not set,  we set it to 5551212.

	       In "full logging" mode,  iWar didn't log timeouts.  Fixed.

	       Added more user submitted banners!


[01/14/2006] - 0.07 Released, "Now with 100% more VoIP!"  - |337Jesus 

               Added support for IAX2.   You can now wardial via the IAX2
	       protocol!  iWar is a full blown IAX2 client!

	       Added sending DTMF tones when using IAX2 mode. (0-9, *, #)

	       Fixed small bugs with command line arg's (output file)

	       The default is no longer DTR hangups,  but plughangup()
	       (+++ATH).  This is pretty generic and supported by everything.
	       If you _can_ use DTR drops,  then use it.  Typically, 
	       Linux supports DTR drops great,  while BSD's don't (probably
	       my fault? ).

	       Fixed problem with changing the volume during dialing. 
	       (iWar would hang up sending a ASCII 13,   but the 
	       modemqueue wouldn't be processed).

	       Fixed major problem with "saving state".  When scanning, 
	       then saving,  then loading,  and saving again,   it would
	       mangle the savestate file.

	       Fixed bug in key stroke marking.  This could cause screwy
	       hangups,  and was also slowing down keyboard response time!
	       This was a issue brought to my attention several times.
               
	       Included patch (snprint/strncpy) from Kevin Anderson.  I made
	       some minor changes,  as his patch used OpenBSD strlcpy/
	       strlcat (changed to strncpy/strncat). 

	       Added "dtrsec" option.   This allows the user to set the
	       amount of time during a DTR drop (might make hanging up
	       work better on some modems).

	       Fixed up the plushangup function to work correctly.  If your
	       modem doesnt support DTR drops,  it should be safe to use
	       the plushangup. 

	       Fixed multiple FreeBSD issues (probably fixed other BSDs).

	       Made the configure.in a little more robust.   For example,
	       added --disable-iax2 and --disable-mysql in the 
	       go ole "./configure" flags. 

	       Fixed strftime() warnings.   These actually lead to seg.
	       faults on the AMD64.  All fixed!

	       More code cleanup.

[12/01/2005] - 0.06 Released, "Protovision,  I have you now...."

               Made the screen redraw (when resizing the screen) into the
	       function - drawinfo()

	       Added more banners!  Thanks to m2mike on binrev.com.
	       If you've got a banner, send it my way!

	       Added "dtrinit" option.  Some modems (USR Couiers) treat DTR
	       hangups like sending a "ATZ".  You can probably disable
	       this in the modem firmware.   With this enabled,  after
	       a hangup/skip/comment/connect,  after DTR drops the
	       modem will be re-initialized. 

	       Added a new terminal window at the bottom of the screen.
	       This allows you to watch iWar as it works.  If it connects
	       to a remote system,  the data will be displayed there. 
	       (Ends up being a nice feature to help debug problems
	       to!).

	       Fixed issue with "Screen full,  Clearing...." (Didn't
	       clear the entire screen all the time).

	       Fixed major problem with "Toneloc" style tone location.  
	       Modem init's and such would cause for a false positive
	       "FOUND TONE!".   It's now set to ignore init string's.
	       Also,  modem did not hang up after tone detection.  This
	       is now done during tone scanning (Toneloc style) by 
	       sending a ATH0.   Since this is the case,  you can't
	       have "dtrinit" and "tone_detection" enabled in the 
	       iwar.conf.  When iWar starts,  it automatically detects
	       if both are set,  and adjusts.   dtrinit isn't needed
	       during tone location as we are sending a ATH0 and not
	       dropping DTR. 

	       Modified the INIT string to support S06 and S07. 
	       This has worked all along,  just DTR on USR's screws
	       it up (hence the dtrinit option in the iwar.conf).

	       Added "TONE" detection.  I have never seen a modem that
	       responds this way, but I've been told they exist.  Might
	       as well support it.  (If you know what type of modem supports
	       this,  please let me know.  I've been told older USR
	       Couriers do this).

	       Fixed a minor issue with seq/random scanning in that it would
	       always miss the last number.  Easy fix.

	       Fixed a major issue where Seq. scanning and loading
	       from a state file would result in incomplete scan.  It
	       didnt verify that the number had already been dialed in the
	       past correctly.

	       Fixed a similar, but seperate issue with random scanning
	       and loading states.  When scanning random,  would result
	       in incomplete scans,  and would also dial the same numbers
	       multiple times.  

	       Fixed problem when saving state that cause the previous
	       filename to be concatenated to current ( pointed out by
	       dominatus on binrev.com)

	       Fixed small problem with screen clearing. 

	       Marking numbers took a bit of time.   At the request of
	       dominatus,  I speed this up a bit. 

	       Added a "ring_timeout" option.  Rather than wait for serial
	       timeout (which might cause you to miss something),  you 
	       can set a timeout between rings.   For example,  if you
	       get two rings,  and not another ring for X number of seconds

iwar-ncurses.c
/*
** Copyright (C) 2005,2006 Softwink,  Inc. <beave@softwink.com>
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 2 of the License, or
** (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
*/


#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "version.h"
#include <curses.h>
#include <string.h>
#include <form.h>
#include <unistd.h>
#include <stdlib.h>
#include "iserial.h"

void nstatus( const char * );
void ntimer( int timer );
void ncount( int ncount);  
void nintro( void );
void npause( int );
void ninfo(const char *, int );  /* 1 = ERROR , 2 = WARN */
void mainscreen( void );
int  nplot(long long , int , int );
void nright(int, int);

char *nsimpleform();
char *nfilename();

void mainscreen( void )
{
	int b;
	int c;
	int maxrow;
	int maxcol;

	getmaxyx(stdscr,maxrow,maxcol); /* Current screen attributes */

	if (maxrow < 24 || maxcol < 80)
 	 {
	 endwin();
	 fprintf(stderr, "This program requires a screen size of 80x24\n");
	 exit(1);
	 }

	init_pair(1, COLOR_WHITE, COLOR_BLUE);
	init_pair(2, COLOR_RED,   COLOR_BLUE);

	init_pair(3, COLOR_WHITE, COLOR_RED);     /* Error messages */
	init_pair(4, COLOR_BLACK, COLOR_YELLOW);  /* Warn message   */
	init_pair(5, COLOR_BLACK, COLOR_WHITE);   /* Intro message  */
	init_pair(6, COLOR_RED,   COLOR_WHITE);   /* Intro border   */


	init_pair(10, COLOR_WHITE, COLOR_BLACK);  /* NO CARRIER */
	init_pair(11, COLOR_GREEN, COLOR_BLACK);  /* CONNECT    */
	init_pair(12, COLOR_YELLOW, COLOR_BLACK); /* BUSY       */
	init_pair(13, COLOR_BLUE,  COLOR_BLACK);  /* VOICE      */
	
	init_pair(14, COLOR_MAGENTA,  COLOR_BLACK);  /* Already scanned */
	init_pair(15, COLOR_CYAN, COLOR_BLACK);      /* Blacklist number */
	init_pair(16, COLOR_RED, COLOR_BLACK);       /* Skipped number */

	attron(COLOR_PAIR(1));
	move(0,0); addch(ACS_ULCORNER);

        for (b = 0; b < maxcol-2; b++) { addch(ACS_HLINE); }
	addch(ACS_URCORNER);

        /* Probably a better way to do this .... */

	for (c=1; c < 9;  c++)
	{
	move(c,0); 
	addch(ACS_VLINE);
		for (b = 0; b < maxcol-2; b++) { addch(' '); } 
	addch(ACS_VLINE);
	}

	attron(COLOR_PAIR(10));

        move(maxrow-8,0); addch(ACS_ULCORNER);
        for (b = 0; b < maxcol-2; b++) { addch(ACS_HLINE); }
        addch(ACS_URCORNER);
        
	move(maxrow-8, maxcol-20); printw("[Terminal Window]");
	for (c=maxrow-7; c < maxrow;  c++)
	        {
	        move(c,0);
	        addch(ACS_VLINE);
		move(c,maxcol-1);
	        addch(ACS_VLINE);
	        }
         
         move(maxrow-1,0); addch(ACS_LLCORNER);
         for (b = 0; b < maxcol-2; b++) { addch(ACS_HLINE); }
         addch(ACS_LRCORNER);

	attron(COLOR_PAIR(1));

	move(1,0); addch(ACS_VLINE); printw(" Port Info       :"); 
	move(2,0); addch(ACS_VLINE); printw(" Start/End Scan  :");
        move(3,0); addch(ACS_VLINE); printw(" Pre/Post Dial   :");
        move(4,0); addch(ACS_VLINE); printw(" Log File        :");
	move(5,0); addch(ACS_VLINE); printw(" Status          :");
	move(6,0); addch(ACS_VLINE); printw(" Idle            :");

	move(8,0);
	addch(ACS_LLCORNER);
	for (b = 0; b < maxcol-2; b++) { addch(ACS_HLINE); }
	addch(ACS_LRCORNER);

	attron(COLOR_PAIR(1));
	move(1, maxcol-23);
	printw("CONNECT      :");
	move(2, maxcol-23);
	printw("NO CARRIER   :");
	move(3, maxcol-23);
	printw("BUSY         :");
	move(4, maxcol-23);
	printw("VOICE        :");
	move(5, maxcol-23);
	printw("TONE/SILENCE :");
        move(6, maxcol-23);
	printw("TIMEOUT      :");
        move(7,maxcol-23);
        printw("Numbers Left : ");


	refresh();   /* Draw! BANG! BANG! */
}

/* "Status:" part of the screen */

void nstatus(const char *nstring)
{
	      init_pair(1, COLOR_WHITE, COLOR_BLUE);
	      attron(COLOR_PAIR(1));
              move(5,20);
	      printw("                                 ");
	      move(5,20);
              printw("%s", nstring);
	      attroff(COLOR_PAIR(1));
              refresh();
}

/* Shows serial timer in seconds */

void ntimer(int ntimer)
{
	     init_pair(1, COLOR_WHITE, COLOR_BLUE);
	     attron(COLOR_PAIR(1));
             move(6,20);
             printw("         ");
             move(6,20);
             printw("%d", ntimer);
             refresh();
}

/* Plot number on the screen.  Color is preset before getting here */

int nplot(long long dialnum, int row,  int col)
{
	    move(row,col);
	    printw("%lld", dialnum);
	    refresh();
	    return 0;
}

/* Some basic statistics */

void nright(int type, int value)
{
int maxcol, maxrow, rrow;
getmaxyx(stdscr,maxrow, maxcol);

if (type==1) rrow=1;    /* CONNECT */
if (type==2) rrow=2;    /* NO CARRIER */
if (type==3) rrow=3;    /* BUSY */
if (type==4) rrow=4;    /* VOICE */
if (type==5) rrow=5;    /* TONE or silence detected */
if (type==6) rrow=6;	/* TIMEOUT */

attron(COLOR_PAIR(1) | A_NORMAL );
move(rrow, maxcol-8);
printw("%d", value);
attroff(COLOR_PAIR(1) | A_NORMAL );
refresh();
}

/* Number of numbers left to dial */

void ncount(int ncount)
{	   
	   int maxrow;
	   int maxcol;
	   getmaxyx(stdscr,maxrow,maxcol);
	   attron(COLOR_PAIR(1));
           move(7, maxcol-8);
	   printw("%d ", ncount);
	   refresh();
}

/* Draws information box in the center of the screen.  Yellow for warning, 
   red for errors (determined by mtype) */

void ninfo(const char *msg, int mtype)
{

WINDOW *info;

int i; int b;
int maxrow;  int maxcol;
getmaxyx(stdscr,maxrow,maxcol);

info = newwin(7,50, (maxrow - 7) / 2, (maxcol - 50) / 2);

if (mtype == 1) { wattrset(info, COLOR_PAIR(3)); }
if (mtype == 2) { wattrset(info, COLOR_PAIR(4)); }

box(info, ACS_VLINE, ACS_HLINE);

/* Fill in box */

for (i=1; i<6; i++)
 { for (b=1; b<49; b++) { mvwaddch(info,i,b, ' '); } }

mvwprintw(info, 3, (50-strlen(msg)) / 2, "%s",  msg);
wrefresh(info);
sleep(1);
delwin(info); 
touchwin(stdscr);
refresh();
}

/* Screen for when the user pauses iWar */

void npause(int ntype)
{
WINDOW *info;

int i; int b;
int maxrow;  int maxcol;
getmaxyx(stdscr,maxrow,maxcol);

info = newwin(7,50, (maxrow - 7) / 2, (maxcol - 50) / 2);

wattrset(info, COLOR_PAIR(4));
box(info, ACS_VLINE, ACS_HLINE);

for (i=1; i<6; i++)
{ for (b=1; b<49; b++) { mvwaddch(info,i,b, ' '); } }

if ( ntype == 0 ) mvwprintw(info, 3, 9, "Paused - Hit any key to continue.");
if ( ntype == 1 ) mvwprintw(info, 3, 9, "Paused and Marked: Any Key Resumes");

wrefresh(info);

while (getch() == -1)
	{
	sleep(1);
	}

delwin(info);
touchwin(stdscr);
refresh();
}

/* Dialog box to get filename to save information */

char *nfilename()
{
FIELD *field[2];
FORM  *form;
WINDOW *form_win;

char buf[255]="";

int key, row, col, maxrow, maxcol;
int i, b;

char tmp[255]="";
char *bufpoint;

getmaxyx(stdscr, maxrow, maxcol);
field[0] = new_field(1,53,0,0,0,0);
field[1] = NULL;

set_field_fore(field[0], COLOR_PAIR(10));
set_field_back(field[0], COLOR_PAIR(10));

form = new_form(field);
scale_form(form, &row, &col);

form_win = newwin( 4,  58, (maxrow - 7)/2, (maxcol - 57)/2);

keypad(form_win, TRUE);

wattron(form_win, COLOR_PAIR(6));
for (i=1; i<8; i++)
{ for (b=1; b<57; b++) { mvwaddch(form_win,i,b, ' '); } }

set_form_win(form, form_win);
set_form_sub(form, derwin(form_win, row, col, 2, 2));
box(form_win, ACS_VLINE, ACS_HLINE);
mvwprintw(form_win, 1, 2, "Enter Filename To Save State To: [Nothing Aborts]", row, col);
post_form(form);
wrefresh(form_win);

while((key = wgetch(form_win)) != 0x0a )
   {
          switch(key)
	            {
                    case KEY_BACKSPACE:
                    form_driver(form, REQ_DEL_PREV);
                    form_driver(form, REQ_DEL_CHAR);

                   /* This is just wrong,  I think.. but works */
                    snprintf(tmp, sizeof(tmp), buf);
                    strncpy(buf, tmp, sizeof(buf));
                    break;
		    
                    /* KEY_BACKSPACE never seems to work for me.  
		    These serve as a "backup" */

                    case 0x08:
                    form_driver(form, REQ_DEL_PREV);
                    form_driver(form, REQ_DEL_CHAR);
                    snprintf(tmp, sizeof(tmp), buf);
                    strncpy(buf, tmp, sizeof(buf));
                    break;
                    case 0x7f:
                    form_driver(form, REQ_DEL_PREV);
                    form_driver(form, REQ_DEL_CHAR);
                    snprintf(tmp, sizeof(tmp), buf);
                    strncpy(buf, tmp, sizeof(buf));
                    break;
                    default:
                    form_driver(form, key);

                   snprintf(tmp, sizeof(tmp), "%c", key);
                   strncat(buf, tmp, sizeof(buf));
                   }
              }

form_driver(form, REQ_VALIDATION);
unpost_form(form);
free_form(form);
free_field(field[0]);
delwin(form_win);
touchwin(stdscr);
refresh();

bufpoint=buf;
return bufpoint;

}


/* Get user input about a number.  This creates a dialog box for the user
   to enter data */

char *nsimpleform()
{

FIELD *field[2];
FORM  *form;
WINDOW *form_win;

int key, row, col, maxrow, maxcol;
int i, b;
char buf[255]="";
char tmp[255]="";
char *bufpoint;

getmaxyx(stdscr, maxrow, maxcol);

field[0] = new_field(4,53,0,0,0,0);
field[1] = NULL;

set_field_fore(field[0], COLOR_PAIR(10));
set_field_back(field[0], COLOR_PAIR(10));

form = new_form(field);
scale_form(form, &row, &col);

form_win = newwin( 8,  58, (maxrow - 7)/2, (maxcol - 57)/2);
keypad(form_win, TRUE);

wattron(form_win, COLOR_PAIR(6));
for (i=1; i<8; i++)
{ for (b=1; b<57; b++) { mvwaddch(form_win,i,b, ' '); } }

set_form_win(form, form_win); 
set_form_sub(form, derwin(form_win, row, col, 2, 2));
box(form_win, ACS_VLINE, ACS_HLINE);
mvwprintw(form_win, 1, 2, "Enter Comment For This Number:", row, col);
post_form(form);
wrefresh(form_win);

while((key = wgetch(form_win)) != 0x0a )
   {
       switch(key)
          {
	   	case KEY_BACKSPACE:
			form_driver(form, REQ_DEL_PREV);
			form_driver(form, REQ_DEL_CHAR);
	
                        snprintf(tmp, sizeof(tmp), buf);
                        strncpy(buf, tmp, sizeof(buf));
			break;

		case 0x08:
			form_driver(form, REQ_DEL_PREV);
                        form_driver(form, REQ_DEL_CHAR);
                        snprintf(tmp, sizeof(tmp), buf);
                        strncpy(buf, tmp, sizeof(buf));
                        break;
		case 0x7f:
			form_driver(form, REQ_DEL_PREV);
                        form_driver(form, REQ_DEL_CHAR);
                        snprintf(tmp, sizeof(tmp), buf);
                        strncpy(buf, tmp, sizeof(buf));
                        break;
		default:
			form_driver(form, key);
			snprintf(tmp, sizeof(tmp), "%c", key);
                        strncat(buf, tmp, sizeof(buf));

			}
	  }

form_driver(form, REQ_VALIDATION);
unpost_form(form);
free_form(form);
free_field(field[0]);
delwin(form_win);
touchwin(stdscr);
refresh();

bufpoint=buf;
return bufpoint;

}

/* Hello.. My name is Da Beave.  I wrote this program,  and I'd like to
   welcome you to it.   That's what this part does :) */

void nintro()
{ 
WINDOW *info;

int i; int b;
int maxrow;  int maxcol;
char msg[50];

getmaxyx(stdscr,maxrow,maxcol);
info = newwin(15,50, (maxrow - 7) / 2, (maxcol - 50) / 2);
wattrset(info, COLOR_PAIR(6));
box(info, ACS_VLINE, ACS_HLINE);

wattrset(info, COLOR_PAIR(5));

for (i=1; i<14; i++)
{ for (b=1; b<49; b++) { mvwaddch(info,i,b, ' '); } }

strncpy(msg, "Intelligent War Dialer  [iWar]", sizeof(msg)); mvwprintw(info, 1, (50-strlen(msg)) / 2, "%s", msg);
strncpy(msg, "http://www.softwink.com/iwar", sizeof(msg));  mvwprintw(info, 2, (50-strlen(msg)) / 2, "%s", msg);
snprintf(msg, sizeof(msg), "Version: %s", VERSION); mvwprintw(info, 3, (50-strlen(msg)) / 2, "%s", msg);
strncpy(msg, "\"Now with 101% more VoIP!\"", sizeof(msg)); mvwprintw(info, 4, (50-strlen(msg)) / 2, "%s", msg);
wattrset(info, COLOR_PAIR(6));
strncpy(msg, "Written By Da Beave", sizeof(msg)); mvwprintw(info, 6, (50-strlen(msg)) / 2, "%s", msg);
wattrset(info, COLOR_PAIR(5));
strncpy(msg, "Released Jan. 16th 2006", sizeof(msg)); mvwprintw(info, 8, (50-strlen(msg)) / 2, "%s", msg);
strncpy(msg, "Contact Information:", sizeof(msg)); mvwprintw(info, 10, (50-strlen(msg)) / 2, "%s", msg);
strncpy(msg, "Email: beave@softwink.com", sizeof(msg)); mvwprintw(info, 11, (50-strlen(msg)) / 2, "%s", msg);
strncpy(msg, "IRC: irc.telephreak.org #telephreak", sizeof(msg)); mvwprintw(info, 12, (50-strlen(msg)) / 2, "%s", msg);

wrefresh(info);
sleep(3);
delwin(info);
touchwin(stdscr);
refresh();
}
